src/System.Private.CoreLib/shared/System/Array.cs

```diff
@@ -893,11 +893,10 @@ public static void ForEach<T>(T[] array, Action<T> action)
         // 
         public IEnumerator GetEnumerator()
```

Comment:
Sharing `GetEnumerator` with CoreCLR is a perf regression for CoreRT.

`SZArrayEnumerator` calls into Array:GetValue in its `get_Current` implementation that is going to duplicate useless checks for range and rank. Can this and `SZArrayEnumerator` be `#define`d out for CoreRT?

Reply:
What about not hacking it for CoreRT but make it better for everyone by introducing `object GetValueNoBoundChecks(int)` which will be in CoreRT/CoreCLR/Mono parts and do only pointer type check and return the value. Looking at the implementation there are many more places where this could be used because the range has been checked

Reply:
Yep, but [I didn't want to force the work as part of the merge](https://github.com/dotnet/coreclr/pull/22321#discussion_r253261146). :)

Reply:
If it was not clear, the suggestion was not to do the improvements as part of this PR but in parallel to this PR in repos where this change is already in.

Reply:
I have missed that `SZArrayEnumerator` path is also different. I will put it under the ifdef as well for now.

I agree that there is goodness from CoreRT to be factored into to the shared enumerator implementation; but the call to `GetValueWithFlattenedIndex_NoErrorCheck` in the enumerator is not quite right:
- The standard for collections is to not have buffer overruns when misused from multiple threads. The CoreRT implementation of the enumerator does not have this robustness. 
- CoreCLR multi-dimm arrays can go up to 4G of elements. The flat index in the shared enumerator needs to be uint.